model/config.go

```diff
@@ -61,34 +44,68 @@ func (k Key) String() string {
 // Object references supplied and returned from this interface should be
 // treated as read-only. Modifying them violates thread-safety.
 type ConfigRegistry interface {
-	// Get retrieves a configuration element, bool indicates existence
-	Get(key Key) (proto.Message, bool)
+	// Get retrieves a configuration element.
+	Get(kind, key string) (config proto.Message, exists bool, revision Revision)
 
-	// List returns objects for a kind in a namespace.
-	// Use namespace "" to list resources across namespaces
-	List(kind string, namespace string) (map[Key]proto.Message, error)
+	// List returns objects by type indexed by the key
+	List(kind string) (map[string]proto.Message, error)
 
 	// Post creates a configuration object. If an object with the same
 	// key already exists, the operation fails with no side effects.
-	Post(key Key, v proto.Message) error
+	Post(kind string, v proto.Message) (Revision, error)
 
 	// Put updates a configuration object in the distributed store.
 	// Put requires that the object has been created.
-	Put(key Key, v proto.Message) error
+	// Revision prevents overriding a value that has been changed
+	// between prior _Get_ and _Put_ operation to achieve optimistic concurrency.
+	Put(kind string, v proto.Message, revision Revision) (Revision, error)
 
 	// Delete removes an object from the distributed store by key.
-	Delete(key Key) error
+	Delete(kind, key string) error
+}
+
+// IstioRegistry is an interface to access config registry using Istio
+// configuration types
+type IstioRegistry interface {
+	// RouteRules lists all routing rules indexed by keys
+	RouteRules() map[string]*proxyconfig.RouteRule
+
+	// RouteRulesBySource selects routing rules by source service instances.
+	// A rule must match at least one of the input service instances since the proxy
+	// does not distinguish between source instances in the request.
+	// The rules are sorted by precedence (high first) in a stable manner.
+	RouteRulesBySource(instances []*ServiceInstance) []*proxyconfig.RouteRule
+
+	// IngressRules lists all ingress rules
+	IngressRules() map[string]*proxyconfig.RouteRule
+
+	// DestinationRules lists all destination rules
+	DestinationRules() []*proxyconfig.DestinationPolicy
+
+	// DestinationPolicy returns a policy for a service version.
+	DestinationPolicy(destination string, tags Tags) *proxyconfig.DestinationVersionPolicy
 }
 
+// Revision is an opaque identifier for auditing updates to the config registry.
```

Comment:
Thanks. I will move the explanations to the code comments. The revision can be used for auditing updates, but we don't place a requirement on the config storage to preserve all records of updates. There should be a separate audit system that logs all mutations together with the revision ID that can work together with the config store.